### ISSUE TITLE
Create an OrtValue from a pytorch tensor on CUDA

### ISSUE THREAD

**Issue by jeshels** [microsoft/onnxruntime#10217]
In python, if I have a numpy array, ```x```,  and want to create an OrtValue on a CUDA device I do the following:
```ortvalue = onnxruntime.OrtValue.ortvalue_from_numpy(x, 'cuda', 0)```

But if I have a torch tensor, ```y```, that is already in CUDA/GPU memory, it seems I have to copy it to CPU memory before creating the OrtValue:
```onnxruntime.OrtValue.ortvalue_from_numpy(y.data.cpu().numpy(), 'cuda', 0)```

Is there a way to create the OrtValue directly from the the torch tensor, such that the tensor doesn't have to be copied to CPU memory and then back to GPU memory?

_Originally posted by @mattroos in https://github.com/microsoft/onnxruntime/discussions/7529_

*Update:* A workaround of providing a GPU tensor to `io_bindings.bind_input()` without explicitly creating an OrtValue would be welcomed as well.

**Comment by pranavsharma** [comment]
Does scenario 2 documented [here](https://onnxruntime.ai/docs/api/python/api_summary.html#iobinding) not work for you? You can fetch the torch data ptr using [this](https://pytorch.org/docs/stable/generated/torch.Tensor.data_ptr.html#torch.Tensor.data_ptr).

**Comment by jeshels** [comment]
Thank you very much!
I'm afraid to admit that I wasn't aware of this documentation, and I tried to figure my way out from the tests.
I successfully passed a `torch` tensor allocated on CUDA gpu device directly to my ONNX model without copying it to the CPU - using `iobiding` and the tensor `data_ptr()` method.

PS. Creating an OrtValue from NumPy array (scenario 2 from the docs) works as expected. I didn't have issues with that.

**Comment by jeshels** [comment]
For future reference, it is also possible to create OrtValue from Torch tensor via dlpack. See  #10286

**Comment by pommedeterresautee** [comment]
and much simpler -> just provide the torch pointer to the output of your model, etc.:
```python
binding = session.io_binding()
...
# test export to Torch tensor directly
# https://onnxruntime.ai/docs/api/python/api_summary.html#iobinding
logit_output = torch.empty((2, 128, 50257), dtype=torch.float32, device='cuda')
binding.bind_output(name=onnx_named_outputs[0], device_type='cuda', device_id=0, element_type=np.float32, shape=tuple(logit_output.shape), buffer_ptr=logit_output.data_ptr())
```

**Comment by farzanehnakhaee70** [comment]
Is there any way to initiate the torch tensor without specifying its shape (I need it for the inputs with dynamic shape)?

**Comment by pommedeterresautee** [comment]
unlike TensorRT, ONNX Runtime has no way to infer output shape from input shape.
So it's up to the user to guess the output shape. In my case I have found that naming input and output axis in the same way to indicate that they should have the same size is a good solution, but i may not work for you.

**Comment by farzanehnakhaee70** [comment]
Thanks a lot for your response.
Is it possible that we allocate the size to the maximum and our input size will be less than the maximum?

**Comment by jeshels** [comment]
As I said, it is possible. See here:
https://github.com/microsoft/onnxruntime/blob/fdce4fa6af437b0b822958ab47b3b8f77f9e14ae/orttraining/orttraining/python/training/ortmodule/_utils.py#L50-L59

**Comment by farzanehnakhaee70** [comment]
So nice. Thanks a lot @jeshels

**Comment by JoachimMoe** [comment]
I have tried reading through some of the discussions posted here, and I am still in doubt as to how to do this correctly with dynamic outputs. For example, I have a RetinaNet instance, and I wish to pre-process my images on the GPU directly without having host-device communication in the InferenceSesssion. Therefore, I can not create an IObinding as the output shapes can change. This solution:

```
def run_with_data_on_device(x):
    session = create_session(MODEL_FILE)

    
    x_ortvalue = ort.OrtValue._ortvalue_to_torch_tensor(x, DEVICE_NAME, DEVICE_INDEX)

    io_binding = session.io_binding()
    io_binding.bind_input(name='images', device_type=x_ortvalue.device_name(), device_id=0, element_type=x.dtype, shape=x_ortvalue.shape(), buffer_ptr=x_ortvalue.data_ptr())
    io_binding.bind_output(name=session.get_outputs()[-1].name, device_type=DEVICE_NAME, device_id=DEVICE_INDEX, element_type=x.dtype, shape=x_ortvalue.shape())
    session.run_with_iobinding(io_binding)

    z = io_binding.get_outputs()

    return z[0]
```

Works wonderfully, but I would wish that I could simply avoid the ```from_numpy``` call and simply pass a Tensor already on device as such:

```
def run_with_torch_tensors_on_device(x: torch.Tensor, CURR_SIZE: int, torch_type: torch.dtype = torch.float) -> torch.Tensor:
    session = create_session(MODEL_FILE)
    binding = session.io_binding()
    x_tensor = x.contiguous()

   ## PROBLEMATIC TENSOR CREATION HERE, PREDICTIONS CHANGE SO NO WAY TO KNOW BEFOREHAND..
   ## torch.zeros(None, None) WOULD BE AWESOME :D 
 z_tensor = torch.zeros((CURR_SIZE,1000), dtype=torch_type, device=DEVICE).contiguous()

 


    binding.bind_input(
        name=session.get_inputs()[0].name,
        device_type=DEVICE_NAME,
        device_id=DEVICE_INDEX,
        element_type=np.float32,
        shape=tuple(x_tensor.shape),
        buffer_ptr=x_tensor.data_ptr())


    binding.bind_output(
        name=session.get_outputs()[-1].name,
        device_type=DEVICE_NAME,
        device_id=DEVICE_INDEX,
        element_type=np.float32,
        shape=tuple(z_tensor.shape),
        buffer_ptr=z_tensor.data_ptr())

    session.run_with_iobinding(binding)

    return z_tensor.squeeze(0)

```

Am I overlooking something?
